021-server-side-rendering/021-server-side-rendering.md

```diff
@@ -0,0 +1,296 @@
+
+The SSR application must set up the success page to call the exchange endpoint. The page can then set session cookie on the SSR domain.
+`
+
+**Problem #3: Solution A - setSession helper method**
+
+A new SDK helper method, called `setSession`, `setToken`, `setSecret`, or `setSessionSecret` to set the session token of future requests.
+
+```js
+import { Client, Account } from "appwrite";
+
+const client = new Client();
+client.setEndpoint("https://cloud.appwrite.io/v1");
+client.setProject("PROJECT_ID");
+
+client.setSession(sessionToken);
+
+const account = new Account(client);
+const currentUser = await account.get();
+```
+
+### New API Endpoints
+
+**PUT /v1/account/sessions/oauth/exchange** - Exchange a temporary OAuth2 token for a session token.
+
+**Request**
+
+| Name   | Type   | Description             |
+| ------ | ------ | ----------------------- |
+| token  | String | Temporary OAuth2 token. |
+| userId | String | User ID.                |
+
+**Response**
+
+Session object.
+
+### Data Structure
+
+**Session Object Additions**
+
+| Name   | Type   | Description    |
+| ------ | ------ | -------------- |
+| secret | String | Session token. |
+
+### Documentation & Content
+
+#### What **docs** would support this feature?
+
+- SDK docs and examples for the new helper method.
+- SDK docs and examples for the new token exchange endpoint.
+
+#### What **tutorials** (text/video) might help developers understand this feature scope, capabilities, and possible use-cases?
+
+- Generic Tutorial for using Appwrite for SSR
+- Tutorial for using Appwrite and Next.js for SSR
+- Tutorial for using Appwrite and SvelteKit for SSR
+
+#### What **demo applications** can help us demonstrate this feature APIs and capabilities?
+
+- Update Almost SSR examples.
+
+### Prior art
+
+[prior-art]: #prior-art
+
+- https://supabase.com/docs/guides/auth/server-side-rendering
+- https://firebase.nuxtjs.org/tutorials/ssr.html
+- https://github.com/gladly-team/next-firebase-auth
+
+### Unresolved questions
+
+[unresolved-questions]: #unresolved-questions
+
+**Problem #1**
+
+1.1. Which solution should we implement? A or B?
+
+**Problem #2**
+
+2.1. Should we remove the undocumented workaround?
+
+2.2. Do we need to add a new endpoint to exchange the temporary token for a session token? Can we reuse the existing magic URL exchange endpoint?
+
+2.3. The Create OAuth2 Session endpoints are not included with Server Side SDKs. Should we include them? If so, server side apps need the Location header to forward to the user. How should we make this accessible from the SDK?
+
+**Problem #3**
+
+3.1. Should the helper method be available in Server Side SDKs? How do we seperate requests that are authenticated with a session token, and requests that are authenticated with an API key?
+
+**General**
+
+4.1. Incorporating some of these SSR changes reduces the distinction between the Server Side and Client Side SDKs. Which SDKs would be recommended for SSR applications?
```

Comment:
Exactly, it gets harder to differentiate. Also, I worry about rate-limiting, which would currently happen using SSR, right?

Reply:
@loks0n let's check where can rate limit get in the way. If our key is email+ip we should be fine. If the abuse key for an endpoint is only for IP it will be useless.

Reply:
Rate limiting is an issue, I've added a section outlining some possible solutions